[FEAT] 관리자 멤버 목록 페이지네이션 도입

jye326 opened the pull request:
# Issue Number
closed #878

## As-Is
<!-- 문제 상황 정의 -->
- 멤버 데이터의 양이 늘어나 관리자 페이지에서 전체 멤버 조회에 걸리는 시간이 증가하였습니다.

## To-Be
<!-- 변경 사항 -->
- 관리자 전용 멤버 조회 로직을 기존 `MemberService`에서 관리자 전용 서비스 레이어로 분리하였습니다.
- 관리자 페이지의 멤버 조회 방식을 페이지 조회로 변경하였습니다.
- 이를 위한 멤버 페이징 조회 API가 추가되었습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## 이슈

- 관리자 페이지 조회 용 api 이므로 디자인보다는 가독성과 직관성이 중요하여 무한 스크롤 방식이 아닌 페이지 넘버링 방식을 선택했습니다.
- 구현 난이도를 고려해 커서 기반이 아닌 Offset 기반 페이징을 도입했습니다. 
  - Offset 기반 페이징은 후반부 데이터를 조회할 때 전반부 데이터의 offset을 스킵 해야 하는 비용이 있습니다. 따라서 O(n)의 시간복잡도를 가집니다.
  - 현재 멤버 데이터의 규모가 크지 않아 (개발 서버 기준 약 3만 건, 10~50ms) 후반부 페이지 조회 시에도 감당할 만 한 수준이라고 판단됩니다.
  - 추후 데이터가 더욱 증가할 시, 검색 기능 추가와 페이징 방식 변경을 고려할 수 있습니다.

jye326 committed:
feat: AdminMemberService 분리 및 관련 컨트롤러 적용

- `AdminMemberService` 신규 추가 및 관리자 회원 조회 로직 분리
- `AdminMemberController`에서 `AdminMemberService`로 서비스 주입 변경

jye326 committed:
feat: 관리자 회원 페이징 조회 기능 추가

- `AdminMemberController`에 회원 페이징 조회 API 추가
- `CustomMemberRepository` 및 구현체 추가하여 페이징 쿼리 적용
- `AdminMemberService`에 페이징 로직 추가 및 기존 메서드 수정
- `PageResult` DTO 추가하여 페이징 결과 반환 지원

jye326 committed:
refactor: AdminMemberService 불필요한 import 제거 및 코드 가독성 개선

- `CustomMemberRepository` import 제거
- `PageResult` 제네릭 타입 선언 간소화

jye326 committed:
feat: AdminMemberService 페이징 조회 테스트 및 로직 개선

- `AdminMemberServiceTest` 추가 및 관리자 회원 페이징 조회 테스트 작성
- `FixtureUtil` 내 전화번호 생성 로직 가독성 개선
- 페이징 조회 쿼리에 정렬 및 조건 로직 추가 (`where`, `orderBy`)
- 멘토, 멘티 테스트 데이터 생성 시 역할과 ID 수정
- 관련 종속성 `@Import` 및 `@MockitoBean` 설정 추가

jye326 committed:
feat: AdminMember 조회 테스트 반환 타입 변경 및 검증 로직 수정

- `AdminMemberIntegrationTest`에서 반환 타입을 `PageResult`로 변경
- 페이징 결과에서 `content()`를 이용한 검증 로직 추가
- 불필요한 주석 제거 및 `Assertions` 호출 대상 수정

jye326 committed:
refactor: AdminMemberService 및 PageResult 개선

- PageResult에 total, totalPages, hasNext 필드 추가로 페이징 정보 확장
- AdminMemberService에서 total, totalPages, hasNext 계산 로직 추가
- AdminMemberServiceTest에 관련 필드 검증 테스트 추가
- 코드 가독성과 유지보수성 향상

kimdevv commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomMemberRepositoryImpl.java:
```diff
@@ -0,0 +1,39 @@
+
+@RequiredArgsConstructor
+public class CustomMemberRepositoryImpl implements CustomMemberRepository {
+
+    private static final QMember member = QMember.member;
+
+    private final JPAQueryFactory jpaQueryFactory;
+
+    @Override
+    public List<Long> findMemberIdsForAdmin(int page, int size) {
+        long offset = (long) (page - 1) * size;
+
+        return jpaQueryFactory.select(member.id)
+                .from(member)
+                .orderBy(member.id.desc())
+                .offset(offset)
+                .limit(size)
+                .fetch();
+    }
+
+    @Override
+    public List<AdminMemberResponse> findMembersByIdsOrdered(List<Long> ids) {
+        return jpaQueryFactory.select(
+                        Projections.constructor(AdminMemberResponse.class,
+                                member.name, member.loginId, member.gender, member.phone.number, member.role)
+                ).from(member)
+                .where(member.id.in(ids))
+                .orderBy(member.id.desc())
+                .fetch();
+    }
```
p3: `findMembersByIdsOrdered()` 메서드 하나만으로 페이지네이션 도입이 가능했을 것 같은데, `findMemberIdsForAdmin()`으로 멤버의 id를 먼저 찾는 이유가 궁금합니다!

jye326 replied:
offset 스킵 비용을 최소화하기 위해서 입니다! offset 기반 페이징은 offset만큼을 읽고 버리는데, Member의 크기가 큰 다른 칼럼들을 읽고 버리면 IO 비용이 커집니다. 따라서 먼저 Id만 읽고, 해당 Id 들만 본문을 읽어 IO를 최적화 합니다!

kimdevv replied:
그렇군요. 접근 방식이 정말 좋네요 👍

kimdevv commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminMemberService.java:
```diff
@@ -0,0 +1,35 @@
+        long total = memberRepository.count();
+        int totalPages = (int)Math.max(1, (total + size - 1) / size);
+        boolean hasNext = responses.size() == size;
+        return new PageResult<>(responses, page, size, total, totalPages, hasNext);
```
p2: `PageResult`의 `size` 필드에는 항상 일정한 값(지금은 20)이 들어갈 것 같아요. 마지막 페이지의 경우 실제로는 20개보다 적은 개수를 반환할 수도 있는데 `size` 값은 항상 20이 될 수 있을 것 같습니다! `responses.size()` 등으로 수정하는 건 어덜까요?

jye326 replied:
좋은데요?

kimdevv commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminMemberService.java:
```diff
@@ -0,0 +1,35 @@
+        List<AdminMemberResponse> responses = memberRepository.findMembersByIdsOrdered(ids);
+        long total = memberRepository.count();
+        int totalPages = (int)Math.max(1, (total + size - 1) / size);
+        boolean hasNext = responses.size() == size;
```
p2: 레코드 개수가 100개 있는 경우에는 마지막 페이지인 5번째 페이지에서도 레코드의 개수가 20개가 나오기 때문에 `hasNext`가 true로 나올 수 있을 것 같아요.
`boolean hasNext = (certificates.size() == PAGE_SIZE_OF_CERTIFICATE) && (page != totalPages);` 등으로 하는 게 좋을 것 같습니다!

jye326 replied:
좋네요! 그런데 당장 필요하지 않은 변수라는 의견이 있어 hasNext를 지웠습니다! 👍

kimdevv commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminMemberService.java:
```diff
@@ -0,0 +1,35 @@
+
+    private final MemberRepository memberRepository;
+
+    public PageResult<AdminMemberResponse> findAllForAdminPaged(Long adminId, int page, int size) {
```
p2: 페이지에 포함될 레코드의 개수(size)는 컨트롤러에서 주입하는 게 아니라 서비스 계층에서 상수 등으로 관리하는 건 어떤가요? 일종의 비즈니스 관련 변수니까요!

jye326 replied:
클라이언트가 페이지에 포함될 레코드 개수를 입력할 수 있도록 지원하기 위함입니다! 10개씩 보기, 20개씩 보기 같은 옵션이 있는 레퍼런스가 떠올랐어요!

kimdevv reviewed (changes requested):
고생하셨습니다~~ 저도 어드민 자격증명 페이지네이션 조회 구현하면서 많이 참고했어요!
커멘트 좀 남겨놨으니 확인 부탁드립니다~~

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/presentation/dto/PageResult.java:
```diff
@@ -0,0 +1,13 @@
+    int size,
+    long total,
+    int totalPages,
+    boolean hasNext
```
p2: 해당 필드(hasNext)는 사용하지 않아서 제거해주면 좋을 것 같아요!

jye326 replied:
다른 페이징 작업하시는 분이 쓰려나 해서 만들었는데 일단 제거했습니다~ 필요하면 추가해서 쓰시면 될 것 같아요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomMemberRepository.java:
```diff
@@ -0,0 +1,10 @@
+public interface CustomMemberRepository {
+
+    List<Long> findMemberIdsForAdmin(int page, int size);
+    List<AdminMemberResponse> findMembersByIdsOrdered(List<Long> ids);
```
p2: 개행 부탁드립니다

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomMemberRepositoryImpl.java:
```diff
@@ -0,0 +1,39 @@
+    public List<AdminMemberResponse> findMembersByIdsOrdered(List<Long> ids) {
+        return jpaQueryFactory.select(
+                        Projections.constructor(AdminMemberResponse.class,
+                                member.name, member.loginId, member.gender, member.phone.number, member.role)
```
p2: 프로젝션 생성자 부분도 개행 부탁드려요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/application/member/repository/MemberRepository.java:
```diff
@@ -1,12 +1,13 @@
 import org.springframework.data.repository.ListCrudRepository;
 
-public interface MemberRepository extends ListCrudRepository<Member, Long> {
+public interface MemberRepository extends ListCrudRepository<Member, Long>, CustomMemberRepository {
```
p3: `CustomMemberRepository`를 상속하신 이유가 있나요? 
지금의 `CustomMemberRepository`는 어드민 전용 쿼리만 있어서 굳이 도메인 레포지토리에서 상속받지 않고 하나의 구현체로서 @Repository를 붙여줘도 될 것 같은데, 이렇게 구현하신 의도가 궁금해요!

jye326 replied:
'리포지토리는 도메인 객체의 저장소를 추상화한 인터페이스'라는 관점에서 같은 Member 도메인 객체를 다루기 때문입니다! 따라서 둘을 완전히 분리하지 않고 페이징 및 관리자 전용 조회 등의 일부 메서드만 인터페이스로 분리하고 기존 `MemberRepository`가 상속 받는 지금의 형태로 구현하였습니다.

ppparkta reviewed (commented):
수고하셨어요! 띠용 덕에 기본적인 페이지네이션 뼈대가 생겨서 통일성 있게 작업할 수 있겠네요 👍 
코멘트 몇 개 남겼는데 확인부탁드립니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomMemberRepositoryImpl.java:
```diff
@@ -0,0 +1,39 @@
+@RequiredArgsConstructor
+public class CustomMemberRepositoryImpl implements CustomMemberRepository {
+
+    private static final QMember member = QMember.member;
```
p2: 여기도 상수 컨벤션 맞춰주세요!

jye326 committed:
refactor: AdminMemberService 페이징 로직 및 PageResult 구조 개선

- `PageResult`에서 불필요한 필드(`hasNext`) 제거 및 구조 단순화
- `AdminMemberService`의 페이징 로직 반환 매개변수 수정
- `CustomMemberRepositoryImpl`의 반환 데이터 정렬 및 가독성 개선
- 사용하지 않는 불필요한 클래스 필드 정리 (`member` static 변수 변경)

jye326 committed:
refactor: AdminMemberServiceTest 검증 로직 단순화

- 테스트 검증에서 불필요한 `hasNext` 필드 검증 제거
- 검증 로직 축소를 통해 코드 가독성 개선

kimdevv reviewed (approved):
고생하셨습니다. 저는 이만 승인합니다~

ppparkta reviewed (approved):
코멘트 확인했습니다. 고생많으셨어요!
